File: lib/aura/memory/metabolizer.rb
# frozen_string_literal: true

module Aura
  module Memory
    class Metabolizer
      def initialize(store:, policy:, summarizer:, metabolism_config: {}, event_bus: nil, registry: nil)
        @store = store
        @policy = policy
        @summarizer = summarizer
        @metabolism_config = metabolism_config
        @event_bus = wrap_event_bus(event_bus)
        @registry = registry
      end

      def run_if_needed
        stats = {
          total_events: 0,
          candidates_for_summary: 0,
          summarized: 0,
          deleted: 0,
          errors: 0
        }

        begin
          stats[:total_events] = @store.count_events

          return stats unless should_metabolize? && stats[:total_events] > recent_events_n

          emit(:metabolism_start, event_count: stats[:total_events], total_chars: @store.total_events_chars)

          old_events = select_old_events
          return stats if old_events.empty?

          retention_result = @policy.apply(old_events)
          stats[:candidates_for_summary] = retention_result[:to_summarize].size

          if retention_result[:to_summarize].any?
            summary = generate_metabolism_summary(retention_result[:to_summarize])
            if summary && !summary.empty?
              @store.insert_summary(content: "Metabolism: Narrative Summary - #{summary}")
              stats[:summarized] = retention_result[:to_summarize].size
              emit(:metabolism_summary, content: summary)
            end
          end

          ids_to_delete = retention_result[:to_delete].map { |e| e["id"] }
          if ids_to_delete.any?
            @store.delete_events(ids_to_delete)
            stats[:deleted] = ids_to_delete.size
            emit(:metabolism_complete, deleted_count: stats[:deleted])
          end
        rescue StandardError => e
          stats[:errors] += 1
          $stderr.puts "[Memory::Metabolizer] Error: #{e.message}"
          $stderr.puts e.backtrace.first(5).join("\n")
        end

        stats
      end
      alias run run_if_needed

      private

      def wrap_event_bus(bus)
        return bus if bus.nil? || bus.is_a?(EventBus)
        EventBus.new(bus)
      end

      def should_metabolize?
        total_chars = @store.total_events_chars
        event_count = @store.count_events
        max_chars = @metabolism_config[:max_chars] || 100_000
        recent_n = recent_events_n

        total_chars > max_chars || event_count > recent_n * 5
      end

      def recent_events_n
        @metabolism_config[:recent_events_n] || 20
      end

      def select_old_events
        keep_recent = recent_events_n
        offset = [keep_recent - 1, 0].max
        @store.fetch_events(offset: offset)
      end

      def generate_metabolism_summary(events)
        return nil unless summarization_enabled?

        max_chars = @metabolism_config.dig(:summarization, :max_chars) || 500
        summary = @summarizer.synthesize(events)
        summary = summary[0, max_chars] if summary && summary.length > max_chars
        summary
      rescue StandardError => e
        $stderr.puts "[Memory::Metabolizer] Summary generation failed: #{e.message}"
        nil
      end

      def summarization_enabled?
        val = @metabolism_config.dig(:summarization, :enabled)
        val.nil? ? true : (val == true)
      end

      def emit(event, data = {})
        @event_bus&.emit(event, data) if @event_bus
      end
    end
  end
end

File: src/core/memory/metabolizer.ts
import type { IEventBus } from '../kernel/interfaces.js';
import type { ToolRegistry } from '../kernel/registry.js';
import { MemoryEventBus } from './eventBus.js';
import type { MemoryPolicy } from './policy.js';
import type { EventRecord as Event, SQLiteStore } from './sqliteStore.js';
import type { MemorySummarizer } from './summarizer.js';

export interface MetabolismConfig {
  max_chars?: number;
  recent_events_n?: number;
  summarization?: {
    enabled?: boolean;
    max_chars?: number;
  };
}

export interface MetabolismResult {
  total_events: number;
  candidates_for_summary: number;
  summarized: number;
  deleted: number;
  errors: number;
}

export class MemoryMetabolizer {
  private store: SQLiteStore;
  private policy: MemoryPolicy;
  private summarizer: MemorySummarizer;
  private metabolismConfig: MetabolismConfig;
  private eventBus?: MemoryEventBus;
  private registry?: ToolRegistry;

  constructor(options: {
    store: SQLiteStore;
    policy: MemoryPolicy;
    summarizer: MemorySummarizer;
    metabolismConfig?: MetabolismConfig;
    eventBus?: IEventBus;
    registry?: ToolRegistry;
  }) {
    this.store = options.store;
    this.policy = options.policy;
    this.summarizer = options.summarizer;
    this.metabolismConfig = options.metabolismConfig || {};
    this.eventBus = options.eventBus
      ? this.wrapEventBus(options.eventBus)
      : undefined;
    this.registry = options.registry;
  }

  public async runIfNeeded(): Promise<MetabolismResult> {
    const stats: MetabolismResult = {
      total_events: 0,
      candidates_for_summary: 0,
      summarized: 0,
      deleted: 0,
      errors: 0,
    };

    try {
      stats.total_events = this.store.countEvents();

      if (
        !this.shouldMetabolize() ||
        stats.total_events <= this.recentEventsN
      ) {
        return stats;
      }

      this.emit('metabolism_start', {
        event_count: stats.total_events,
        total_chars: this.store.totalEventsChars(),
      });

      const oldEvents = this.selectOldEvents();
      if (oldEvents.length === 0) {
        return stats;
      }

      const retentionResult = this.policy.apply(oldEvents);
      stats.candidates_for_summary = retentionResult.to_summarize.length;

      let summaryGenerated = false;
      if (retentionResult.to_summarize.length > 0) {
        const summary = await this.generateMetabolismSummary(
          retentionResult.to_summarize,
        );
        if (summary?.trim()) {
          let latestEventId: number | null = null;
          try {
            const latestEvents = this.store.fetchEvents({ limit: 1 });
            if (latestEvents.length > 0) {
              latestEventId = latestEvents[0].id;
            }
          } catch (_e) {}

          this.store.insertSummary({
            content: `Metabolism: Narrative Summary - ${summary}`,
            source_event_id: latestEventId,
          });
          stats.summarized = retentionResult.to_summarize.length;
          summaryGenerated = true;
          this.emit('metabolism_summary', { content: summary });
        }
      }

      // Only delete summarization candidates if they were successfully summarized,
      // or if they are purely non-summarizable ephemeral entries.
      const toDelete = retentionResult.to_delete.filter((event) => {
        const needsSummary = this.policy.shouldSummarize(event);
        return !needsSummary || summaryGenerated;
      });

      const idsToDelete = toDelete.map((e: Event) => e.id).filter(Boolean);
      if (idsToDelete.length > 0) {
        this.store.deleteEvents(idsToDelete);
        stats.deleted = idsToDelete.length;
        this.emit('metabolism_complete', { deleted_count: stats.deleted });
      }
    } catch (e: unknown) {
      stats.errors += 1;
      console.warn(`[Memory::Metabolizer] Error: ${(e as Error).message}`);
    }

    return stats;
  }

  public async run(): Promise<MetabolismResult> {
    return this.runIfNeeded();
  }

  private wrapEventBus(bus: IEventBus): MemoryEventBus {
    if (bus instanceof MemoryEventBus) {
      return bus;
    }
    return new MemoryEventBus(bus);
  }

  private shouldMetabolize(): boolean {
    const totalChars = this.store.totalEventsChars();
    const eventCount = this.store.countEvents();
    const maxChars = this.metabolismConfig.max_chars || 100000;
    const recentN = this.recentEventsN;

    return totalChars > maxChars || eventCount > recentN * 5;
  }

  private get recentEventsN(): number {
    return this.metabolismConfig.recent_events_n || 20;
  }

  private selectOldEvents(): Event[] {
    const allEvents = this.store.fetchEvents();
    const oldEvents: Event[] = [];
    const reversed = [...allEvents].reverse();
    const counts: Record<string, number> = {};

    for (const event of reversed) {
      const phase = event.phase;
      const tool = event.tool;
      const policy = this.policy.getRetentionPolicy(phase, tool);

      if (policy.permanent === true) {
        continue;
      }

      const maxSteps = policy.max_steps ?? this.recentEventsN;
      const hasManifestPolicy = tool ? (typeof this.policy.getManifestRetention === 'function' && this.policy.getManifestRetention(tool) !== null) : false;
      const key = hasManifestPolicy ? `${phase}:${tool}` : phase;

      counts[key] = (counts[key] || 0) + 1;

      if (counts[key] > maxSteps) {
        oldEvents.push(event);
      }
    }

    return oldEvents.reverse();
  }

  private async generateMetabolismSummary(
    events: Event[],
  ): Promise<string | null> {
    if (!this.summarizationEnabled) {
      return null;
    }

    const maxChars = this.metabolismConfig.summarization?.max_chars || 500;
    let summary = await this.summarizer.synthesize(events);
    if (summary && summary.length > maxChars) {
      summary = summary.substring(0, maxChars);
    }
    return summary;
  }

  private get summarizationEnabled(): boolean {
    const val = this.metabolismConfig.summarization?.enabled;
    return val === undefined || val === true;
  }

  private emit(event: string, data: Record<string, unknown> = {}): void {
    this.eventBus?.emit(event, data);
  }
}
